### ISSUE TITLE
verifyAtLeastOneInteractionOn needed

### ISSUE THREAD

**Issue by GoogleCodeExporter** [dlsgusrn7577/mockito#239]
```
I like Mockito because the Mockito help me write decoupled tests to 
implementation. 

In this point of view, I have an idea to improve Mockito. Sometimes, I want to 
only verify whether there is an interaction(s) on mock object regardless of the 
interaction of any method.

For example, following is a simple code to do something.

if (type = MONEY) {
  command.execute();
} else {
  otherCommand.execute);
}

A test may look like this.

@Test
public void commandShouldBeExecuted_WhenTypeIsMoney() {
  sut.doIt();

  verify(command).execute();
}

Now, the test has known the method 'execute' of 'command'. That can be 
unnecessary coupling on some situations. I think this problem can be resolved 
by providing new API 'verifyAtLeastOneInteractionOn' which check whether there 
is at least one interaction on mock object. 

@Test
public void commandShouldBeExecuted_WhenTypeIsMoney() {
  sut.doIt();

  verifyAtLeastOneInteractionOn(command);
}

Thanks for reading.
```

Original issue reported on code.google.com by `minslo...@gmail.com` on 6 Jan 2011 at 9:36

**Comment by GoogleCodeExporter** [comment]
```
Sorry for my mistake. This issue is enhancement, not defect.
```

Original comment by `minslo...@gmail.com` on 6 Jan 2011 at 9:38

**Comment by GoogleCodeExporter** [comment]
```
Thanks for feedback!

Frankly, I'm not sure about this feature. If I don't care what method was 
called why should I bother testing that in first place? 

Usually, when we're unsure about a feature we seek for real-life examples. Does 
your proposal come from the example you faced in real world?
```

Original comment by `szcze...@gmail.com` on 6 Jan 2011 at 8:12

**Comment by GoogleCodeExporter** [comment]
Original comment by `szcze...@gmail.com` on 6 Jan 2011 at 8:12
- Added labels: **Type-Enhancement**
- Removed labels: **Type-Defect**

**Comment by GoogleCodeExporter** [comment]
```
Thanks for your reply. Following is my real example.

First, our faced problem is the stability of tests. Kent Beck explains this 
concern in his article(http://www.threeriversinstitute.org/WhereToTest.html). 
The more tests know about test targets, the more stability problem is 
increased. As a result, the cost of test sustaining is increased. In our 
development environment, this problem is not a light issue.

Following is a real example and a related test piece.

public class ServiceStopper {
    Stop completeStop;
    Stop readOnlyStop;
    Notifier readOnlyNotifier;

    public enum StopType {
        COMPLETE_STOP, READONLY_STOP;
    }

    public void stop(StopType stopType, String reason) {
        if (stopType == null) {
            throw new IllegalArgumentException("StopType cannot be null.");
        }

        if (isEmpty(reason)) {
            throw new IllegalArgumentException("Reason cannot be empty.");
        }

        if (stopType == StopType.COMPLETE_STOP) {
            completeStop.stopNow();
        } else if (stopType == StopType.READONLY_STOP) {
            readOnlyStop.stopNow();
            readOnlyNotifier.notifyToRelatedServices(createStopMessageBy(reason));
        }
    }
}

@RunWith(MockitoJUnitRunner.class)
public final class ServiceStopperTest {
    private final ServiceStopper target = new ServiceStopper();
    private final StopType completeStopType = StopType.COMPLETE_STOP;
    private final StopType readOnlyStopType = StopType.READONLY_STOP;
    private final Boolean success = Boolean.TRUE;
    private final Boolean fail = Boolean.FALSE;

    @Mock
    private Stop completeStop;
    @Mock
    private Stop readOnlyStop;
    @Mock
    private Notifier readOnlyNotifier;

    @Test
    public void serviceShouldBeStopped_WhenNeedingCompleteStop() {
        target.stop(completeStopType, "Completely stop test");

        verify(completeStop).stopNow();
    }
        ...
}

In this case, interaction on 'completeStop' is only 'stopNow()' method. So, if 
there is at least one interaction, it means actually there was an interaction 
of 'stopNow()' and the test should be passed. 

Of course, if the interaction of mock is more than one, this idea can be not 
applied. However, I have been seeing many cases to apply this idea. 

In a nutshell, I would like that tests are decoupled from implementations. 

Thanks for reading.









```

Original comment by `minslo...@gmail.com` on 7 Jan 2011 at 1:52

**Comment by GoogleCodeExporter** [comment]
```
So basically, you would prefer your test to look like:

  @Test
  public void serviceShouldBeStopped_WhenNeedingCompleteStop() {
    //when
    target.stop(completeStopType, "Completely stop test");
    //then
    verifyAtLeastOneInteractionOn(completeStop);
  }

??

Let me paraphrase your test:

"when 'target' is told to stop completely,
some interaction should happen with 'completeStop' collaborator, 
I don't really care what interaction, be it toString() or stop() or 
drinkFiveBeers()"

Your test should describe the behavior of the system under test. Is above the 
behavior you want to describe?
```

Original comment by `szcze...@gmail.com` on 8 Jan 2011 at 12:33

**Comment by GoogleCodeExporter** [comment]
```
Exactly! Above test, you wrote, is what I want. 
The test describes "When 'stopNow' of SUT is invoked, completeStop should has 
at least one interaction" 

Of course, above test does not describe "what is interaction of completeStop?". 
If this approach obscures to reveal the intention of the test and test 
readability is the first value to users, this way may be not appropriate.  On 
the contrary, if users prefer to improve stability through hiding the internal 
of SUT from the test, this approach may be more valuable.

As you know, there are many situations in real world. If frameworks or 
libraries should provide the choice for users, it may be good.

Thanks for your feedback and reading.


```

Original comment by `minslo...@gmail.com` on 8 Jan 2011 at 3:57

**Comment by GoogleCodeExporter** [comment]
```
I'm afraid this test does not make any sense to me :) If I don't care what 
method was called I should not bother testing this at all.

I suggest you to write to Kent / write to the TDD mailing list he hangs / or 
comment on his blog. Show him your case ask him what would he do. Seriously, we 
suppose to be a supporting community. Alternatively, you can ask mockito 
mailing list to get more opinions. Personally, I fail see the case for this 
feature. Although I very much appreciate tests that are only focused on the 
essential behavior, hiding away irrelevancies (this actually a motto for 
mockito).
```

Original comment by `szcze...@gmail.com` on 8 Jan 2011 at 5:14

**Comment by GoogleCodeExporter** [comment]
```
Hi,

I completely agree with Szczepan.

I also read this article of Kent Beck a while ago. There is a difference 
between high level test (like integration tests) and unit tests.
What Kent Beck probably meant by stability is not relevant in unit testing.

In your specific case would there be any reason to your test to fail if the 
behaviour I specified, while your production code is actually working, I don't 
think so. It's precisely what stability is about (in the sense Kent Beck 
expressed).
```

Original comment by `brice.du...@gmail.com` on 17 Jan 2011 at 2:27

**Comment by GoogleCodeExporter** [comment]
```
I do it this way :

{code}
assertThat(
new 
MockUtil().getMockHandler(mockedService).getInvocationContainer().getInvocations
(), hasSize(greaterThanOrEqualTo(1))
{code}
```

Original comment by `biethb@gmail.com` on 27 Sep 2011 at 12:46
